client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -108,6 +108,17 @@ public void readerOffline(String readerId, Position lastPosition) {
 
     @Override
     public CompletableFuture<Checkpoint> initiateCheckpoint(String checkpointName, ScheduledExecutorService backgroundExecutor) {
+
+        synchronizer.fetchUpdates();
```

Comment:
This is not the correct way to use the synchronizer. You should apply a conditional update to add the checkpoint and abort if there are too many. Using it in this way doesn't guarantee anything. Instead this should be a modification to the call on line 122.

Reply:
Will `fetchUpdates()` not guarantee to reflect the latest snapshot of the distributed state? Why do we need to add the checkpoint and fail later?